app/concepts/jwt_api_entreprise/contract/admin_create.rb

```diff
@@ -1,10 +1,54 @@
 module JwtApiEntreprise::Contract
-  AdminCreate = Dry::Validation.Schema do
-    required(:roles) { filled? { each { str? } } }
-    required(:user_id).filled(:str?)
-    required(:subject).maybe(:str?)
+  class AdminCreate < Reform::Form
+    property :user_id
+    property :subject
 
-    # TODO Now create contacts when creating a token
-    # collection :contacts, form: Contact::Contract::Upsert, populate_if_empty: Contact
+    validation do
+      configure do
+        config.messages_file = Rails.root
+          .join('config/dry_validation_errors.yml').to_s
+
+        def user_exists?(uid)
+          User.exists?(uid)
+        end
+      end
+
+      required(:user_id).filled(:str?, :user_exists?)
+      required(:subject).filled(:str?)
+
+      # TODO Add a high level validation rule here to ensure at least one
+      # business and one tech contact is provided in the payload
+      required(:contacts).filled
+
+      # TODO Add a rule to validate the format of the :roles data to ensure
+      # the populator won't fail (format should be [{ code: '...' }, { ... }])
+      required(:roles).filled
+    end
+
+    collection :contacts,
+      form: Contact::Contract::Upsert,
+      populate_if_empty: Contact
+
+    collection :roles, populate_if_empty: :populate_roles_from_code do
+      property :code
+
+      validation with: { form: true } do
+        configure do
+          config.messages_file = Rails.root
+            .join('config/dry_validation_errors.yml').to_s
+
+          def role_exists?(code)
+            form.model.persisted?
+          end
+        end
+
+        required(:code).filled(:role_exists?)
+      end
+    end
+
+
+    def populate_roles_from_code(options)
+      Role.find_by_code(options[:fragment][:code]) or Role.new
```

Comment:
Je suis assez confus sur cette ligne, tu peux m'expliquer ?...

Reply:
ActiveRecord: `find_by_elephant(foo)` c'est de la méta prog ça fait : `find_by(elephant: foo)`

Reply:
Et concernant les options en paramètre et surtout le `fragment`, on le devine certainement ici il s'agit du "morceau" de la payload passé en paramètre associé au formulaire (ici c'est un formulaire "nesté" pour les rôles que l'on veut attachés aux jetons). 

C'est en vrai vraiment balèze ce que fait la lib, ici elle gère des créations de liens en *Many To Many* entre les jetons et les rôles, avec toutes les validations qui vont bien. Les modèles et les liens entre eux sont créés tout seul en base lorsqu'on appelle la macro *Persist*.

Reply:
Il faut utiliser si possible utiliser || au lieu de or, pour le souci suivant : https://stackoverflow.com/questions/2083112/difference-between-or-and-in-ruby

Bonus points car le or délimite moins fortement a la lecture les 2 parties de l'expression

Reply:
Je vais quand même mettre un || j'aime pas trop les soucis qui peuvent être causés par l'utilisation du or effectivement.